File: control_center/models/proxy.go
package models

import "time"

// ProxySession = autorisation éphémère d'accès à un reverse-proxy applicatif
// (JupyterLab / code-server). Émise par un endpoint authentifié au Bearer, elle est
// référencée par un cookie HttpOnly que le navigateur renvoie automatiquement sur
// l'iframe, les liens et les WebSockets — là où le token JS ne peut pas voyager.
//
// La VM cible (TargetIP/TargetPort) est résolue CÔTÉ SERVEUR au moment de l'émission :
// le client ne choisit jamais une IP. C'est la défense anti-IDOR du dispositif.
type ProxySession struct {
	ID         string `gorm:"primaryKey"` // valeur opaque du cookie (aléatoire, non devinable)
	Email      string `gorm:"index"`      // identité qui a ouvert la session
	Kind       string // "jupyter" | "vscode"
	PoolID     string // serverpool_id (pour matcher le chemin du proxy)
	OwnerID    string // user_id propriétaire du pool
	Target     string // identité de la VM cible (email élève, ou owner pour la VM instructeur)
	VMID       string `gorm:"index"` // UUID de la VM (= chemin du proxy, URL-safe ; évite l'email dans l'URL)
	TargetIP   string // IP résolue côté serveur — jamais fournie par le client
	TargetPort int    // port résolu (8888 Jupyter, 8443 code-server RW, 8444 code-server RO)
	Mode       string // "read" | "write"
	ExpiresAt  time.Time
	CreatedAt  time.Time `gorm:"autoCreateTime"`
}

// VscodeGrant = partage explicite, par une cible, de l'accès à SON VS Code (code-server).
// Un autre utilisateur présente (cible + mot de passe) ; si un grant valide existe, le
// proxy ouvre une ProxySession vers la VM de la cible dans le mode autorisé.
//
// Le prof n'a pas besoin de grant : son rôle (staff + supervision du pool) suffit.
type VscodeGrant struct {
	ID           uint      `gorm:"primaryKey;autoIncrement"`
	PoolID       string    `gorm:"index"` // serverpool_id du pool partagé
	OwnerID      string    // user_id propriétaire du pool
	Target       string    `gorm:"index"` // identité (email) de l'élève qui partage SA VM
	PasswordHash string    // bcrypt du mot de passe de partage
	Mode         string    // "read" | "write"
	ExpiresAt    time.Time // expiration du partage
	CreatedAt    time.Time `gorm:"autoCreateTime"`
}

File: frontend/src/lib/proxy.ts
import { apiFetch } from './api';

// Accès aux applications des VMs (JupyterLab, code-server) via le reverse-proxy HTTPS du
// control center — jamais l'IP directe de la VM (qui reste non exposée).
//
// Le navigateur ne peut pas porter le Bearer token sur une iframe / un nouvel onglet /
// un WebSocket. On demande donc d'abord une « session de proxy » (POST authentifié) qui
// pose un cookie HttpOnly, puis on ouvre l'URL renvoyée : le cookie voyage tout seul.

export type ProxyKind = 'jupyter' | 'vscode';
export type ProxyMode = 'read' | 'write';

export interface ProxyOpened {
  url: string;     // base du proxy (origin-relative), à ouvrir / mettre en iframe src
  mode: ProxyMode;
  target: string;
}

// openProxySession ouvre (et autorise côté serveur) une session de proxy vers une VM.
//   kind   : 'jupyter' | 'vscode'
//   poolId : serverpool_id ; ownerId : user_id propriétaire du pool
//   target : 'self' (sa VM) | 'instructor' | <email élève> (réservé au staff)
//   mode   : 'read' | 'write' (vscode)
// Renvoie l'URL à ouvrir, ou lève une erreur lisible (403/503…).
export async function openProxySession(
  kind: ProxyKind,
  poolId: string,
  ownerId: string,
  target: string = 'self',
  mode: ProxyMode = 'write',
): Promise<ProxyOpened> {
  const res = await apiFetch('/api/proxy-session', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ kind, pool_id: poolId, owner_id: ownerId, target, mode }),
  });
  if (!res.ok) {
    throw new Error((await res.text()) || `proxy-session ${res.status}`);
  }
  return res.json();
}

// rejoindre le VS Code d'un binôme via (cible + mot de passe). Pose le cookie vscode et
// renvoie l'URL à ouvrir, dans le mode autorisé par le grant.
export async function joinVscode(
  poolId: string,
  ownerId: string,
  target: string,
  password: string,
): Promise<ProxyOpened> {
  const res = await apiFetch('/api/vscode-grant/join', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ pool_id: poolId, owner_id: ownerId, target, password }),
  });
  if (!res.ok) {
    throw new Error((await res.text()) || `join ${res.status}`);
  }
  return res.json();
}

// partager SON VS Code : crée un grant (mode + mot de passe + expiration).
export async function shareVscode(
  poolId: string,
  ownerId: string,
  mode: ProxyMode,
  password: string,
  ttlHours = 24,
): Promise<{ ok: boolean; target: string; mode: ProxyMode; expires_at: string }> {
  const res = await apiFetch('/api/vscode-grant', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ pool_id: poolId, owner_id: ownerId, mode, password, ttl_hours: ttlHours }),
  });
  if (!res.ok) {
    throw new Error((await res.text()) || `share ${res.status}`);
  }
  return res.json();
}

// ouvrir l'URL d'un proxy dans un nouvel onglet (Jupyter / VS Code plein écran).
export function openInNewTab(url: string) {
  window.open(url, '_blank', 'noopener');
}
